// @ergosign/storybook-addon-pseudo-states-react/dist/share/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PseudoStatesDefaultPrefixAlternative = exports.PseudoStatesDefaultPrefix = exports.AttributesStatesInputDefault = exports.AttributesStatesDefault = exports.PseudoStatesDefault = exports.COLUMN = exports.ROW = exports.Orientation = exports.HOST_CLASS = exports.PROPERTY = exports.PermutationStatesObjectType = exports.READONLY = exports.DISABLED = exports.AttributeStatesEnum = exports.VISITED = exports.FOCUS_WITHIN = exports.FOCUS = exports.ACTIVE = exports.HOVER = exports.PseudoStateEnum = void 0;
var PseudoStateEnum;
(function (PseudoStateEnum) {
    PseudoStateEnum["HOVER"] = "hover";
    PseudoStateEnum["ACTIVE"] = "active";
    PseudoStateEnum["FOCUS"] = "focus";
    PseudoStateEnum["FOCUS_WITHIN"] = "focus-within";
    PseudoStateEnum["VISITED"] = "visited";
})(PseudoStateEnum = exports.PseudoStateEnum || (exports.PseudoStateEnum = {}));
exports.HOVER = PseudoStateEnum.HOVER, exports.ACTIVE = PseudoStateEnum.ACTIVE, exports.FOCUS = PseudoStateEnum.FOCUS, exports.FOCUS_WITHIN = PseudoStateEnum.FOCUS_WITHIN, exports.VISITED = PseudoStateEnum.VISITED;
var AttributeStatesEnum;
(function (AttributeStatesEnum) {
    AttributeStatesEnum["DISABLED"] = "disabled";
    AttributeStatesEnum["ENABLED"] = "enabled";
    AttributeStatesEnum["READONLY"] = "readonly";
})(AttributeStatesEnum = exports.AttributeStatesEnum || (exports.AttributeStatesEnum = {}));
exports.DISABLED = AttributeStatesEnum.DISABLED, exports.READONLY = AttributeStatesEnum.READONLY;
var PermutationStatesObjectType;
(function (PermutationStatesObjectType) {
    PermutationStatesObjectType["PROPERTY"] = "property";
    PermutationStatesObjectType["HOST_CLASS"] = "host-class";
})(PermutationStatesObjectType = exports.PermutationStatesObjectType || (exports.PermutationStatesObjectType = {}));
exports.PROPERTY = PermutationStatesObjectType.PROPERTY, exports.HOST_CLASS = PermutationStatesObjectType.HOST_CLASS;
var Orientation;
(function (Orientation) {
    Orientation[Orientation["ROW"] = 0] = "ROW";
    Orientation[Orientation["COLUMN"] = 1] = "COLUMN";
})(Orientation = exports.Orientation || (exports.Orientation = {}));
exports.ROW = Orientation.ROW, exports.COLUMN = Orientation.COLUMN;
exports.PseudoStatesDefault = [exports.FOCUS, exports.HOVER, exports.ACTIVE];
exports.AttributesStatesDefault = [
    { attr: exports.DISABLED, value: true },
];
exports.AttributesStatesInputDefault = [
    { attr: exports.DISABLED, value: true },
    { attr: exports.READONLY, value: true },
];
exports.PseudoStatesDefaultPrefix = ':';
exports.PseudoStatesDefaultPrefixAlternative = 'pseudo-states--';
