// @ergosign/storybook-theme/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ergosignTheme = void 0;
var theming_1 = require("@storybook/theming");
exports.ergosignTheme = (0, theming_1.create)({
    base: 'light',
    colorPrimary: '#009de0',
    colorSecondary: '#d7007f',
    // UI
    appBg: '#fff',
    appContentBg: '#fff',
    appBorderColor: '#fff',
    appBorderRadius: 0,
    // Typography
    fontBase: 'Interstate, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif',
    fontCode: '"Fira Code", Menlo, monospace',
    // Text colors
    textColor: '#000',
    textInverseColor: 'rgba(255,255,255,0.9)',
    // Toolbar default and active colors
    barTextColor: 'rgba(255,255,255,0.8)',
    barSelectedColor: '#fff',
    barBg: '#009de0',
    // Form colors
    inputBg: '#fff',
    inputBorder: '#c0c0c0',
    inputTextColor: '#000',
    inputBorderRadius: 4,
    brandTitle: 'Ergosign UX Library',
    brandUrl: 'https://www.ergosign.de',
    brandImage: 'https://www.ergosign.de/assets/img/ergosign-logo.svg'
});
